### ISSUE TITLE
`parse.typed(...)` or some kind of typed parse method

### ISSUE THREAD

**Issue by mmkal** [colinhacks/zod#1748]
It'd be nice to have zod help the developer max out their chances of parsing a valid object, when it's manually constructed. Example:

```ts
const User = z.object({
  name: z.string().regex(/^\w+ \w+$/),
  age: z.number(),
})

User.parse({
  name: 'Bob',
  birthdate: new Date('1970-01-01'),
})
```

The above code compiles, but we have enough information to _know_ that it'll throw a ZodError at runtime. It'd be nice if there was a method like

```ts
User.parse.typed({
  name: 'Bob',
  birthdate: new Date('1970-01-01'),
})
```

Where the expected input type was `Input` (from the `ZodType<Output, Def, Input>`) rather than `unknown`. So the above code would error because the developer used `birthdate` instead of `age`.

This would be especially useful for types with a custom `.refine` or `.regex` method or similar where it could be misleading to use _just_ a type declaration:

```ts
const user: z.infer<typeof User> = {
  name: "Robert'; -- DROP TABLE Students; --"
  age: 40,
}
```

The above will compile, and implies that the `const user` is a validated `User` instance, but it isn't because `name` doesn't match the regex. The current options are to do the above, which is safe-looking at compile time, but unsafe at runtime, or to use `.parse(...)` which is unsafe at compile time and safe-ish at runtime (but will throw errors).

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const userSchema = z.object( {
    name: z.string().regex( /^\w+ \w+$/ ),
    age: z.number(),
} )
type User = z.infer<typeof userSchema>

userSchema.parse( {
    name: 'Bob',
    birthdate: new Date( '1970-01-01' ),
} ) // runtime error, but no compile time error

userSchema.parse( {
    name: 'Bob',
    birthdate: new Date( '1970-01-01' ),
} as User ) // runtime and compile time error
```

**Comment by mmkal** [comment]
Not exactly, because 1) as-casting isn't really safe, since it allows for missing properties and 2) it involves having a type reference for the schema, or creating one manually via `z.infer<...>`.

Re 1) this would compile fine:

```ts
userSchema.parse({
  name: 'Bob',
  // no age!
} as User)
```

But maybe on latest typescript versions `satisfies` would do the trick:

```ts
userSchema.parse({
  name: 'Bob',
  // no age!
} satisfies User) // compile time error
```

For 2) though, it's quite cumbersome when you don't already happen to have a type reference for the schema, e.g.:

```ts
getAPISchema().lookupProcedure('user/create').getInputSchema().parse({
  name: 'Bob',
} satisfies z.infer<???>)
```

The above code would need to be refactored to extract the schema into a variable, `import {z} from 'zod'`, and then use `z.infer<...>`. Those are all barriers on _top_ of the temptation to just use `.parse(...)` which accepts `unknown` so will compile just fine - much less discoverable and much less likely to be used.

**Comment by JacobWeisenburger** [comment]
Would `typedParse()` work for you?

```ts
const typedParse = <Schema extends z.ZodSchema> (
    schema: Schema, data: z.infer<Schema>
): z.infer<Schema> => schema.parse( data )

const userSchema = z.object( {
    name: z.string().regex( /^\w+ \w+$/ ),
    age: z.number(),
} )

console.log(
    typedParse( userSchema, {
        name: 'Bob',
        birthdate: new Date( '1970-01-01' ),
    } ) // runtime and compile time error
)

typedParse(
    loadAPISchemaInfo().lookupProcedure( 'user/create' ).getInputSchema(),
    {
        name: 'Bob',
    }
)
```

**Comment by JacobWeisenburger** [comment]
I don't think that this will ever be added to Zod, because it's my understanding that parse is meant to take something that is unknown and turn it into something that is known at compile time. So when you are trying to parse an object that is statically in the code base, that's something that is already known at compile time. In example code we do this all the time for learning/teaching purposes, but I don't know that I have ever done that in a real app. In real apps I use Zod for checking user input or a response from an api, which are things that can't be known at compile time. I hope this makes sense. Please let me know if you have questions or if I am misunderstanding what you are talking about.

**Comment by mmkal** [comment]
Yes it's a slightly different use case than validating API inputs, and I usually want this when using refinement types, or regex strings, or similar, which perform runtime validation which is _not_ reflected at compile time. And yes that helper function does the trick for most cases - what I'm asking is whether it could become part of zod. There are some edge cases like .transform types with a different input typearg which it'd be sensible to solve in one official place.

**Comment by sneko** [comment]
Just posted https://github.com/colinhacks/zod/issues/1892 that looks finally really the same than what @mmkal is looking for.

I totally agree with him on implementing it inside Zod. Since Typescript no longer exist at runtime, helping people in writing safer code at compilation helps.

In my case I write fixtures and Storybook stories with mocked data and it's sad Zod cannot tell me my object is wrong when it's missing 90% of the expected props. For sure it's not a magic silver bullet since Typescript wouldn't help with custom validation (min/max/regex...) but at least it helps on having the right properties, and the almost expected property types.

**Comment by JacobWeisenburger** [comment]
Another way you can solve this problem:

```ts
type TypedParseSchema<Schema extends z.ZodSchema> = Omit<Schema, 'parse'> & {
    parse ( data: z.infer<Schema> ): z.infer<Schema>
}
const useTypedParse = <Schema extends z.ZodSchema>
    ( schema: Schema ): TypedParseSchema<Schema> => schema

const schemaWithTypedParse = useTypedParse( z.object( {
    email: z.string(),
} ) )

schemaWithTypedParse.parse( 'hello' )
//                          ^^^^^^^
// Argument of type 'string' is not assignable to parameter of type '{ email: string; }'.
```

@mmkal and @sneko 
Would this work for you?

**Comment by sneko** [comment]
Yes it would. It's a bit sad to have to make "parser helpers" for each schema. Do you still think this should be kept outside the `zod` library?

**Comment by JacobWeisenburger** [comment]
> It's a bit sad to have to make "parser helpers" for each schema

What do you mean "for each schema"? The way I wrote it above can be reused on any schema. So you only need 1 `useTypedParse`.

---

> Do you still think this should be kept outside the `zod` library?

I don't know if it should be added in or not. I feel like this is up to @colinhacks.

**Comment by JacobWeisenburger** [comment]
I made a package that has a few utilities for zod and typed parse is one of them. Let me know if it helps.
https://github.com/JacobWeisenburger/zod_utilz#usetypedparsers

**Comment by sneko** [comment]
> What do you mean "for each schema"? The way I wrote it above can be reused on any schema. So you only need 1 useTypedParse.

By making "helpers" I mean if I have `UserSchema`, `CarSchema`... for each one I need to do:
```
const schemaWithXxxTypedParse = useTypedParse(XxxSchema)
```

(but I agree that's a decent workaround for now, even if I need to write it everywhere I have a schema.

**Comment by mmkal** [comment]
I'm inclined to agree - absolutely, that helper works, but if the use case is valid enough for it to exist in a util library, then I'd argue it'd make sense to exist in zod too. It doesn't really add any meaningful maintenance burden, I don't think, since the implementation is just the same as the untyped `parse`. A util library is less discoverable, requires separate installation and importing, might break with new versions etc. etc.

**Comment by JacobWeisenburger** [comment]
I totally agree. But I don't feel like I have the authority to add a feature like this to someone else's library. I am totally fine with maintaining it on my util library because I have the authority to do so.

**Comment by TomKaltz** [comment]
https://github.com/JacobWeisenburger/zod_utilz#usetypedparsers should be part of zod. Very helpful!  @colinhacks

**Comment by wbobeirne** [comment]
I just wanted to voice my support for this. In addition to validating user inputs to my API, I'm using Zod to format responses from my API and I have tripped over this a few times where the type that I'm passing in obviously won't parse, like a `Promise` if I mistakenly forgot to `await` something, or passing in a `T | null` or `T | undefined` where I should have first checked if the value was there. While this isn't the main use-case for Zod, this feels like a really low-cost method to implement that would be useful for cases where you have semi-trusted input you want to parse.

**Comment by mmkal** [comment]
@colinhacks now that zod is seeing more activity recently, what do you think?

**Comment by wbobeirne** [comment]
Decided to put my money where my mouth is and put up https://github.com/colinhacks/zod/pull/3629

**Comment by colinhacks** [comment]
I'd argue it's fundamentally unsound to have a strongly-typed data structure whose type signature you can't trust...so I don't want a prominent happy-path API for supporting this. Plus there's already a language level affordance for this: satisfies. 

```ts
import { z } from "zod";

const myInput = { ... };
const mySchema = z.object({ ... });

// combo
mySchema.parse(myInput satisfies z.infer<typeof mySchema>);
```

It's actually pretty elegant imo, and I'd much rather point people towards language features than add APIs, especially for such a niche use case.

**Comment by willo-icon** [comment]
Satisfies is definitely pretty good for this, that's a great point. You're right about the un-soundness of matching the type exactly, but I'd still like to consider if there's a way to avoid cases where we could be sure that at a type level what you're passing in couldn't possibly be correct. For instance, if I'm passing in a `Promise`, or if my schema is an object but I'm passing in a `string` (like I forgot to `JSON.parse` something before running it through my schema validator.)

This is definitely in nice-to-have territory, but I've run into a few cases where I previously had been able to rely on compile time checks for these kinds of dumb mistakes that I now can only have certainty by testing at runtime.

**Comment by mmkal** [comment]
Pasting my reply on Xtwitter here for posterity!

---

I'm someone who'd want this as the default in most of my codebases! I may be in the minority - but the reason I want it is the same reason `satisfies z.infer<MySchema>` isn't quite what I want - and it's in part that it's for people I work with, not just for me.

In the somewhat-rare-but-not-that-rare cases where we're constructing some kind of object manually and want to make sure it fully conforms to a complex schema and apply any transforms deep in the schema - I want people to get red squigglies when they do `Person.parse({firstName, lastName})` instead of `Person.parse({name: firstName + ' ' + lastName})`. I want them to get the squigglies even when they're not thinking about how they can make sure they get squigglies - in fact, *especially* when they're not thinking about squigglies. If they're careful enough to add `satisfies z.infer<...>` and be thinking about squigglies - they will probably get the shape of the object they're parsing right anyway. Making people write `as never` to opt out instead will trigger people to think "I know what I'm doing, this is ok", and remind code reviewers to sanity check that thought.

I can think of a handful of mini-outages from this kind of thing (especially if I include io-ts days, it had the same default behaviour). There was always an easy to spot dumb mistake/typo - and of course zod ensured the program technically functioned as expected at runtime - bad data was passed to .parse, so we threw a helpful error. But execution short-circuited and the system didn't do what we really wanted it to.

Re your first point - you can trust the signature! Just as much as you can today. It's only an input type. Today, you can pass anything to .parse and it always returns what it says it will, or throws. That's still the case, the compiler just helps you not pass stuff that it knows ahead of time will throw - I'd argue it's more sound.

Finally, I agree about the happy path. What I love about zod is it's value prop is *immediate* and almost anyone with any level of experience can just look at it and start using it successfully. I would only want this to be tucked away under some kind of "Advanced" section of the docs, especially since it uses and advanced TypeScript feature. I would expect the vast majority of zod users not to know about this (even the ones on my team).

Finally finally, thank you again for taking the time to continue to read and reply to this stuff - I might keep lobbing thoughts your way. And I'm actually very encouraged by you not saying "hey good idea let's go do that" easily. I still like *my* ideas and hope you come around. But projects can easily collapse under their own weight if maintainers run enthusiastically with everything, so it's a good thing. And it shows you're not using chatgpt to reply to me. So thanks for that too.

**Comment by colinhacks** [comment]
This is now achievable with the new `.decode()` methods in Zod 4.1, which were added in conjunction with the Codec API: https://colinhacks.com/essays/introducing-zod-codecs
